staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -457,7 +463,10 @@ func (c *Cacher) startCaching(stopChannel <-chan struct{}) {
 		}
 	}()
 
+	// We pre-terminate as many watchers as possible without holding
+	// watchcache lock.
 	c.terminateAllWatchers()
```

Comment:
so we were already terminating watchers here... what watchers were being established between this point and a successful list resulting in a replace call and what was problematic with them? I thought watchers were blocking here when the cache wasn't ready:

https://github.com/kubernetes/kubernetes/blob/6bce018b36fc26f9b002f730421866ff3e8b919e/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L524

I'm not seeing the problem with leaving newly established watchers blocked there until `c.ready.set(true)` gets called, and then proceeding with them...

Reply:
Let's talk offline about it.

Reply:
the answer is that a call to Watch() that is establishing a new watcher could be in this section of code, where it has already passed the `c.ready.wait` check, but has not yet registered the new watcher in a way that `terminateAllWatchers()` would catch and remove:

https://github.com/kubernetes/kubernetes/blob/f90643435ef606afbd156490933b200479c34e2d/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L524-L623

Reply:
it might be helpful to expand this comment with something like:
```
Watchers currently being established might not be terminated by this call, so we need to call again under lock.
Watchers we terminate here who try to re-establish themselves will block until the ready signal.
```

Reply:
after stewing on this more, I think it would be better to prevent adding watchers we know have stale state in the first place, rather than allowing them to be added, maybe delivering some initial queued events, then terminating them

By tracking the readiness generations, I think we can do this with a change like this instead:

```diff
diff --git a/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go b/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go
index 5679b6adf22..b15ddb65711 100644
--- a/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go
+++ b/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go
@@ -530,7 +530,8 @@ func (c *Cacher) Watch(ctx context.Context, key string, opts storage.ListOptions
 		return nil, err
 	}
 
-	if err := c.ready.wait(ctx); err != nil {
+	readyGeneration, err := c.ready.waitGeneration(ctx)
+	if err != nil {
 		return nil, errors.NewServiceUnavailable(err.Error())
 	}
 
@@ -623,9 +624,17 @@ func (c *Cacher) Watch(ctx context.Context, key string, opts storage.ListOptions
 		return newErrWatcher(err), nil
 	}
 
+	addedWatcher := false
 	func() {
 		c.Lock()
 		defer c.Unlock()
+
+		if generation, ok := c.ready.checkGeneration(); !ok || generation != readyGeneration {
+			// We went unready or are on a different generation, avoid adding
+			return
+		}
+		addedWatcher = true
+
 		// Update watcher.forget function once we can compute it.
 		watcher.forget = forgetWatcher(c, watcher, c.watcherIdx, scope, triggerValue, triggerSupported)
 		// Update the bookMarkAfterResourceVersion
@@ -639,6 +648,11 @@ func (c *Cacher) Watch(ctx context.Context, key string, opts storage.ListOptions
 		c.watcherIdx++
 	}()
 
+	if !addedWatcher {
+		// ... close watcher? with or without error?
+		return
+	}
+
 	// Some events might have already been bufferred and not yet processed
 	// by the cacher. By passing the resource version after the initEvents,
 	// we ensure that hose will not be delivered to the watcher for the
diff --git a/staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go b/staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go
index 47e03fe9e27..25e64274645 100644
--- a/staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go
+++ b/staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go
@@ -39,6 +39,7 @@ const (
 //	└---------------------------┘
 type ready struct {
 	state       status        // represent the state of the variable
+	generation  int           // represent the number of times we have transition to ready
 	lock        sync.RWMutex  // protect the state variable
 	restartLock sync.Mutex    // protect the transition from ready to pending where the channel is recreated
 	waitCh      chan struct{} // blocks until is ready or stopped
@@ -60,11 +61,17 @@ func (r *ready) done() chan struct{} {
 
 // wait blocks until it is Ready or Stopped, it returns an error if is Stopped.
 func (r *ready) wait(ctx context.Context) error {
+	_, err := waitGeneration(ctx)
+	return err
+}
+
+// waitGeneration blocks until it is Ready or Stopped, it returns an error if is Stopped, or the generation if ready
+func (r *ready) waitGeneration(ctx context.Context) (int, error) {
 	for {
 		// r.done() only blocks if state is Pending
 		select {
 		case <-ctx.Done():
-			return ctx.Err()
+			return 0, ctx.Err()
 		case <-r.done():
 		}
 
@@ -79,14 +86,15 @@ func (r *ready) wait(ctx context.Context) error {
 			// waiting despite the state moved back to Pending.
 			r.lock.RUnlock()
 		case Ready:
+			generation := r.generation
 			r.lock.RUnlock()
-			return nil
+			return generation, nil
 		case Stopped:
 			r.lock.RUnlock()
-			return fmt.Errorf("apiserver cacher is stopped")
+			return 0, fmt.Errorf("apiserver cacher is stopped")
 		default:
 			r.lock.RUnlock()
-			return fmt.Errorf("unexpected apiserver cache state: %v", r.state)
+			return 0, fmt.Errorf("unexpected apiserver cache state: %v", r.state)
 		}
 	}
 }
@@ -98,6 +106,13 @@ func (r *ready) check() bool {
 	return r.state == Ready
 }
 
+// check returns the current generation and whether it is currently Ready
+func (r *ready) checkGeneration() (int, bool) {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+	return r.generation, r.state == Ready
+}
+
 // set the state to Pending (false) or Ready (true), it does not have effect if the state is Stopped.
 func (r *ready) set(ok bool) {
 	r.lock.Lock()
@@ -107,6 +122,7 @@ func (r *ready) set(ok bool) {
 	}
 	if ok && r.state == Pending {
 		r.state = Ready
+		r.generation++
 		select {
 		case <-r.waitCh:
 		default:
```

I'd probably separate the buffer index and max(initialEventRV) fixes into a separate PR... those are a lot more straightforward and I'd want to be able to backport those independently